### PR TITLE
feat: function name hints for UDFs

### LINKED ISSUE #9392: Add ScalarUDFs in missing function hints / suggested errors
### Is your feature request related to a problem or challenge?

Part of https://github.com/apache/arrow-datafusion/issues/8045

Today there is a nice error message with a hint if a user issues an unknown function

For example, tryng `cos2` will have datafusion suggest `cos`
```
DataFusion CLI v36.0.0
❯ select cos2(1);
Error during planning: Invalid function 'cos2'.
Did you mean 'cos'?
```

However, UDFs are never suggested

For example, the function `abs` has been moved to a ScalarUDF, so now the error message does not suggest this.

```sql
DataFusion CLI v36.0.0
❯ select abs2(1);
Error during planning: Invalid function 'abs2'.
Did you mean 'acos'?
```

but in earlier versions of DataFusion where `abs` was a built in function, it is correctly suggested:

```sql
DataFusion CLI v32.0.0
❯ select abs2(1);
Error during planning: Invalid function 'abs2'.
Did you mean 'abs'?
```

### Describe the solution you'd like

I would like UDFs to be included in the suggestsed the error message

### Describe alternatives you've considered

Here is where the suggestion is created: https://github.com/apache/arrow-datafusion/blob/cf69d38e80448ec7402f622d2adcfc6edf0b387e/datafusion/sql/src/expr/function.rs#L214-L215

I think it will be a fairly straightforward exercise to extend the list in https://github.com/apache/arrow-datafusion/blob/cf69d38e80448ec7402f622d2adcfc6edf0b387e/datafusion/expr/src/function.rs#L80-L94 to 

### Additional context

This came up in the context of https://github.com/apache/arrow-datafusion/pull/9388, https://github.com/apache/arrow-datafusion/pull/9388#discussion_r1506662726

### PR DESCRIPTION
**Pull request by SteveLauC**
## Which issue does this PR close?

Closes #9392.

### CODE REVIEW

File: datafusion/sql/src/planner.rs [hunk 1]
```diff
@@ -85,6 +85,10 @@ pub trait ContextProvider {
 
     /// Get configuration options
     fn options(&self) -> &ConfigOptions;
+
+    fn udfs(&self) -> HashMap<String, Arc<ScalarUDF>>;
```
Comment (alamb) [inline datafusion/sql/src/planner.rs:89]: This sort of exposes details of the implementation (e.g. a `HashMap`)

What about potentially just returning the names following the model of CatalogProvider or SchemaProvider: https://docs.rs/datafusion/latest/datafusion/catalog/schema/trait.SchemaProvider.html#tymethod.table_names


Something like

```suggestion
    /// returns all udf names
    fn udf_names(&self) -> Vec<&str>
```

(or maybe Strings)

**Review by alamb — COMMENTED** [review]
Thanks @SteveLauC  -- this is looking very cool

File: datafusion/sql/src/planner.rs (hunk 1 above)
Comment (SteveLauC) [reply datafusion/sql/src/planner.rs:89]: Nice! I love consistency! Will do it.

**Review by alamb — COMMENTED** [review]
I think this PR also needs tests -- however, I think when you merge this branch up from main, you should have to update at least one test in a sqllogictest file, so you probably don't have to write anything new

File: datafusion/sql/src/expr/function.rs [hunk 2]
```diff
@@ -20,20 +20,63 @@ use arrow_schema::DataType;
...
+pub fn suggest_valid_function(
+    input_function_name: &str,
+    is_window_func: bool,
+    ctx: &dyn ContextProvider,
+) -> String {
+    let valid_funcs = if is_window_func {
+        // All aggregate functions and builtin window functions
+        AggregateFunction::iter()
+            .map(|func| func.to_string())
+            .chain(BuiltInWindowFunction::iter().map(|func| func.to_string()))
```
Comment (Omega359) [inline datafusion/sql/src/expr/function.rs:49]: Would we want to add the ctx.udafs() and ctx.udwfs() here as well?

**Comment by alamb** [comment]
Hi @SteveLauC  -- how is this PR going? Anything we can help with?

File: datafusion/sql/src/planner.rs (hunk 1 above)
Comment (SteveLauC) [reply datafusion/sql/src/planner.rs:89]: Do you want this function to return a `Vec<&str>` or `Vec<String>`, the suggested change uses `Vec<&str>`, but the `table_names()` function in the referenced link uses `Vec<String>`

File: datafusion/sql/src/expr/function.rs (hunk 2 above)
Comment (SteveLauC) [reply datafusion/sql/src/expr/function.rs]: done

**Review by alamb — APPROVED** [review]
Thank you @SteveLauC  -- I tried this out and it worked great

```
❯ select abs2(1);
Error during planning: Invalid function 'abs2'.
Did you mean 'abs'?
❯
```

I also merged up from main and added a test for the functionality (undid the change fro https://github.com/apache/arrow-datafusion/pull/9388)

File: datafusion/sqllogictest/test_files/functions.slt [hunk 3]
```diff
@@ -483,7 +483,7 @@ statement error Did you mean 'arrow_typeof'?
 SELECT arrowtypeof(v1) from test;
 
 # Scalar function
-statement error Invalid function 'to_timestamps_second'
+statement error Did you mean 'to_timestamp_seconds'?
```
Comment (alamb) [inline datafusion/sqllogictest/test_files/functions.slt:486]: Updated test here

**Comment by alamb** [comment]
Thanks again @SteveLauC

**Pull request merged**